### ISSUE TITLE
Cannot find C library archives after recompilation

### ISSUE THREAD

**Issue by linzhp** [bazelbuild/rules_go#2653]
After recompiling a `go_test` target, C library fails to link.

### What version of rules_go are you using?

0.23.10

### What version of gazelle are you using?

0.22.0

### What version of Bazel are you using?

3.5.0

### Does this issue reproduce with the latest releases of all the above?

This issue cannot reproduce in rules_go 0.24. So we may need to cherry pick some commits into the 0.23 releases

### What operating system and processor architecture are you using?

Darwin

### What did you do?

```bash
$ git clone git@github.com:linzhp/bazel_examples.git
$ cd bazel_examples
$ bazel build recompile:go_default_test
```

### What did you expect to see?

Build completed successfully

### What did you see instead?

```
ERROR: /Users/zplin/gocode/src/github.com/linzhp/bazel_examples/recompile/BUILD.bazel:10:8: GoLink recompile/go_default_test_/go_default_test failed (Exit 1) builder failed: error executing command bazel-out/host/bin/external/go_sdk/builder link -sdk external/go_sdk -installsuffix darwin_amd64 -arc ... (remaining 21 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox
external/go_sdk/pkg/tool/darwin_amd64/link: running external/local_config_cc/cc_wrapper.sh failed: exit status 1
clang: error: no such file or directory: 'bazel-out/darwin-fastbuild/bin/external/icu4c/libicu-base.a'
clang: error: no such file or directory: 'bazel-out/darwin-fastbuild/bin/external/icu4c/libicudata.a'

link: error running subcommand: exit status 2
Target //recompile:go_default_test failed to build
Use --verbose_failures to see the command lines of failed build steps.
```

**Comment by jayconrod** [comment]
It looks like this build passes with `v0.23.0` but not `v0.23.10`. It starts failing at c69d159acddf.

I think #2625 needs to be backported. I haven't done a bugfix release since that landed, so that should probably happen soon.

**Comment by linzhp** [comment]
Fixed in  0.23.11
